### DISCUSSION TITLE
... sleep with Micropython

### DISCUSSION THREAD

**Discussion by ewok2** [micropython/micropython#12889] [RP2040 / Pico] [answered]
I have a Pico W which will be powered by a battery and will send data each xx minutes with wifi/socket.

To limit the power needed I would like to low the intensity allways except when transmitting data through wifi.
I have a look at the lightsleep, deepsleep, picosleep.

What I think I have understand : 
- Picosleep seem's the best one (arround 1/2 mA) but need to modify the C code in the firmeware UF2. The "last version" of the ufé precompiled is old (2021) and did not include so much stadard library
=> So if I want to use it I need to compile my "own" firmeware.

- deepsleep need some more IO manipulation on GPIO to cut the wifi power
- lightsleep is in fact almost like sleep on power consomption side

All the post I found on these subject seem's "old"
Do I have the right view ?

What is the 2023 best solution to be arround 2mA consumption ?

Thanks

**Comment by davefes** [comment]
My next "deepsleep" attempt with the PICO is using a SN74HC74, following the article:
https://github.com/bablokb/pcb-pico-en-control
with a DS3231 RTC for wake-up.

**Reply by ewok2** [reply]
Thanks but I would like to avoid HW solution.
Some mA is ok for me no need to go to micro.

**Comment by ewok2** [comment]
Some news on my research on different sleep.
Il am facing with some difficulty with the Intensity mesurement of the Pico W
I try the folowing in loop : 
-5V input power going to a 10 Ohm resistor then to the PICO W pin  power.
The code of the Pico is simple : 
- flash led red
- get a DHT22 sensor values
- get two I2C INA219 values
- connect to wifi
- flash blue led
- go to sleep (5s)

I plug an osciloscope each side of the 10 Ohms resistor to get the "delta voltage" between CH1 and CH2 and so the Intensity value.

It works if I only set "wlan.active(True)" at begining and "wlan.active(False)"
If after "wlan.active(True)" I try a "wlan.connect(secrets['ssid'], secrets['pw'])" the behaviour seem's strange: 
- if htere is the 10 Ohm resistor It does not works, the loop crash.
- if there is no resistor (a wire instead) the loop works and the Pico get an IP

Any idea of what happend. How can I mesure the intensity with Wifi connected?

![Screenshot 2023-11-10 at 17 29 07](https://github.com/micropython/micropython/assets/3373331/2946ba22-7ce6-41b4-b278-af79a610f7d6)

**Reply by robert-hh** [reply, marked as answer]
The current draw can be several 100mA when WiFi starts. So 10Ohm is too large. Use 1Ohm or less.

**Reply by davefes** [reply]
A 1000uF electro after the 1 Ohm might help with the WiFi surge current affecting behaviour.

**Reply by ewok2** [reply]
thanks for help, indeed with a 1 Ohm it works!
For information below the result mesured with sleep / lightsleep / deepsleep : 

![Screenshot 2023-11-11 at 16 17 41](https://github.com/micropython/micropython/assets/3373331/a77b538e-f620-4704-aa50-a2bea822205c)

seem's deepsleep does not have a better value???

And my code below : 

```
import network, rp2
import machine
import time
from secrets import secrets
import dht 
import ina219
import socket

ina = machine.I2C(0, scl=machine.Pin(1), sda=machine.Pin(0))
 
inaBatt = ina219.INA219(ina, addr=64)
inaSolar = ina219.INA219(ina, addr=69)
inaBatt.set_calibration_16V_400mA()
inaSolar.set_calibration_32V_1A()

ledRed = machine.Pin(2,machine.Pin.OUT)
dht22 = dht.DHT22(machine.Pin(4))
deepSleepTime = 50*1000

hostname = '192.168.yy.xx'
portnum = 'zzzzz'

rp2.country('FR')

errorDht22  = 1
errorSolar  = 2
errorBatt   = 3
errorWater  = 4
errorSocket = 5

socketMessage = {}

def wait_for_wifi():
    # Wait for connect or fail
    max_wait = 10
    while max_wait > 0:
        if wlan.status() < 0 or wlan.status() >= 3:
            break
        max_wait -= 1
        time.sleep(1)

    # Handle connection error
    if wlan.status() != 3:
        return wlan.status()
    else:
        status = wlan.ifconfig()
        if serialConnect:
            logPrint('Wifi connected with ip = ' + status[0])
        return wlan.status()

def flash_led(nb, led):
    for i in range(nb*2):
        led.toggle()
        time.sleep_ms(50)
    led.off()


def errorNumSet(number, errorVal):
    errorVal = errorVal | (1 << number)
    return errorVal

def errorNumReset(number, errorVal):
    errorVal = errorVal & (255 ^ (1 << number))
    return errorVal

# Inital setup and connection
wlan = network.WLAN(network.STA_IF)

socketMessage["aquaErrorNum"] = 0

while True:  
    wlan.active(True)
    time.sleep(1)
    wlan.connect(secrets['ssid'], secrets['pw'])
    while (wait_for_wifi() != 3):
        time.sleep(1)
        wlan.connect(secrets['ssid'], secrets['pw'])
        
    try:
        dht22.measure()
        socketMessage["dht22AquaTemp"] = dht22.temperature()
        socketMessage["dht22AquaHum"] = dht22.humidity()
        socketMessage["dht22AquaValid"] = 1
        socketMessage["aquaErrorNum"] =  errorNumReset(errorDht22,socketMessage["aquaErrorNum"])

    except OSError as e:
        socketMessage["aquaErrorNum"] = errorNumSet(errorDht22,socketMessage["aquaErrorNum"])
        socketMessage["dht22AquaValid"] = 0

    try:
        socketMessage["aquaIntBatt"] = inaBatt.current
        socketMessage["aquaVoltBatt"] = inaBatt.bus_voltage
        socketMessage["aquaBattValid"] = 1
        socketMessage["aquaErrorNum"] = errorNumReset(errorBatt,socketMessage["aquaErrorNum"])

    except OSError as e:
        socketMessage["aquaErrorNum"] = errorNumSet(errorBatt,socketMessage["aquaErrorNum"])
        socketMessage["aquaBattValid"] = 0

    try:
        socketMessage["aquaIntSolar"] = inaSolar.current
        socketMessage["aquaVoltSolar"] = inaSolar.bus_voltage
        socketMessage["aquaSolarValid"] = 1
        socketMessage["aquaErrorNum"] = errorNumReset(errorSolar,socketMessage["aquaErrorNum"])

    except OSError as e:
        socketMessage["aquaErrorNum"] = errorNumSet(errorSolar,socketMessage["aquaErrorNum"])
        socketMessage["aquaSolarValid"] = 0

        
    try:
        homeSocket = socket.socket()
        logPrint('try to connect to raspberry socket')
        homeSocket.settimeout(10)
        homeSocket.connect(socket.getaddrinfo(hostname, portnum, socket.SOCK_STREAM)[0][-1])
        homeSocket.send(str(socketMessage).encode())
        homeSocket.close()
        socketMessage["aquaErrorNum"] = errorNumReset(errorSocket,socketMessage["aquaErrorNum"])
    except OSError as e:
        socketMessage["aquaErrorNum"] = errorNumSet(errorSocket,socketMessage["aquaErrorNum"])

    flash_led(2, ledRed)

    wlan.disconnect()
    time.sleep(1)
    wlan.active(False)
    time.sleep(1)
    wlan.deinit()
    time.sleep(1)

    time.sleep(deepSleepTime/1000)
    #machine.lightsleep(deepSleepTime)
    #machine.deepsleep(deepSleepTime)

```

**Reply by davefes** [reply]
My struggles:
https://forums.raspberrypi.com/viewtopic.php?t=357183&sid=2fccabeb55ff043bee5b9ad832c0f286

**Reply by ewok2** [reply]
Yes I have also seen that lightsleep or deepsleep does not works with Thony shell but I do not need this. 
I can live with sleep when testing and passing without usb connection when using it.
